### DISCUSSION TITLE
Trying to delete a replicated bucket will leave it in a broken state

### DISCUSSION THREAD

**Discussion by krisdaniels** [minio/minio#12403] [Q&A]
When trying to delete a bucket in clustered active active sync mode without deleting the replication rules first,
the bucket will end up in a broken state in the sense that there is no way to remove it at all.
Even after removing what is left of the replication rules the bucket can still not be deleted,
mc gives the following error : The specified method is not allowed against this resource.
The only way I was able to recover from this was to rebuild the test cluster.

## Expected Behavior
A delete of a bucket should not leave the bucket in a corrupted state.

## Steps to Reproduce (for bugs)

Start 2 clusters
```
for i in {1..4}; do docker run --name host1$i --rm -d --network minio -e MINIO_REGION_NAME="region1" -e MINIO_ROOT_USER="admin1" -e MINIO_ROOT_PASSWORD="region1psw" -v ~/minio/region1/host$i:/disks -p 901$i:9000 minio/minio server http://host1{1...4}/disks/data{1...4}; done
for i in {1..4}; do docker run --name host2$i --rm -d --network minio -e MINIO_REGION_NAME="region2" -e MINIO_ROOT_USER="admin2" -e MINIO_ROOT_PASSWORD="region2psw" -v ~/minio/region2/host$i:/disks -p 902$i:9000 minio/minio server http://host2{1...4}/disks/data{1...4}; done
```
Start mc docker

```
docker run --rm -ti --network minio --entrypoint=/bin/sh minio/mc
```

Set the aliases
```
mc alias set region1 http://host11:9000 admin1 region1psw
mc alias set region2 http://host21:9000 admin2 region2psw
```

Create replication users, policies and aliases
```
curl -L https://docs.min.io/minio/baremetal/examples/ReplicationAdminPolicy.json -o ReplicationAdminPolicy.json
mc admin policy add region1 ReplicationAdminPolicy ./ReplicationAdminPolicy.json
mc admin user add region1 region1ReplicationAdmin region1adminpsw
mc admin policy set region1 ReplicationAdminPolicy user=region1ReplicationAdmin

mc admin policy add region2 ReplicationAdminPolicy ReplicationAdminPolicy.json
mc admin user add region2 region2ReplicationAdmin region2adminpsw
mc admin policy set region2 ReplicationAdminPolicy user=region2ReplicationAdmin

curl -L https://docs.min.io/minio/baremetal/examples/ReplicationRemoteUserPolicy.json -o ReplicationRemoteUserPolicy.json
mc admin policy add region1 ReplicationRemoteUserPolicy ReplicationRemoteUserPolicy.json
mc admin user add region1 region1ReplicationRemoteUser region1userpsw
mc admin policy set region1 ReplicationRemoteUserPolicy user=region1ReplicationRemoteUser

mc admin policy add region2 ReplicationRemoteUserPolicy ReplicationRemoteUserPolicy.json
mc admin user add region2 region2ReplicationRemoteUser region2userpsw
mc admin policy set region2 ReplicationRemoteUserPolicy user=region2ReplicationRemoteUser

mc alias set region1Replication http://host11:9000 region1ReplicationAdmin region1adminpsw
mc alias set region2Replication http://host21:9000 region2ReplicationAdmin region2adminpsw
```
Create bucket
```
mc mb -l region1/bucket-1
mc mb -l region2/bucket-1
```
Set 2 way replication on bucket
```
mc admin bucket remote add region1Replication/bucket-1 http://region2ReplicationRemoteUser:region2userpsw@host21:9000/bucket-1 --service "replication" --sync
mc admin bucket remote add region2Replication/bucket-1 http://region1ReplicationRemoteUser:region1userpsw@host11:9000/bucket-1 --service "replication" --sync

curl -L https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64 -o jq
chmod +x ./jq
mc replicate add region1Replication/bucket-1 --remote-bucket bucket-1 --arn "$(mc admin bucket remote ls region1Replication/bucket-1 --json | ./jq -r '.RemoteARN')" --replicate "delete,delete-marker"
mc replicate add region2Replication/bucket-1 --remote-bucket bucket-1 --arn "$(mc admin bucket remote ls region2Replication/bucket-1 --json | ./jq -r '.RemoteARN')" --replicate "delete,delete-marker"
```
Copy a file and check it exists
```
mc cp ./jq region1/bucket-1
mc ls region2/bucket-1 --versions
mc ls region1/bucket-1 --versions
```

Try to delete the bucket
```
sh-4.4# mc rb region1/bucket-1
mc: <ERROR> `region1/bucket-1` is not empty. Retry this command with ‘--force’ flag if you want to remove `region1/bucket-1` and all its contents 
sh-4.4# mc rb region1/bucket-1 --force
mc: <ERROR> Failed to remove `region1/bucket-1`. The bucket you tried to delete is not empty. You must delete all versions in the bucket.
sh-4.4# mc rb region1/bucket-1 --force
mc: <ERROR> Failed to remove `region1/bucket-1`. The specified bucket does not exist.
or 
Removed `region1/bucket-1` successfully.
depending on how fast the commands are retried

sh-4.4# mc rb region2/bucket-1     
mc: <ERROR> `region2/bucket-1` is not empty. Retry this command with ‘--force’ flag if you want to remove `region2/bucket-1` and all its contents 
sh-4.4# mc rb region2/bucket-1 --force
mc: <ERROR> Failed to remove `region2/bucket-1`. The specified method is not allowed against this resource.
sh-4.4# mc rb region2/bucket-1 --force
mc: <ERROR> Failed to remove `region2/bucket-1`. The specified method is not allowed against this resource.
```

At this point there is no way to remove region2/bucket-1

Tried removing anything that was left of the replication
```
mc replicate rm --all --force region1Replication/bucket-1
mc admin bucket remote rm --arn "$(mc admin bucket remote ls region1Replication/bucket-1 --json | ./jq -r '.RemoteARN')" region1Replication/bucket-1

mc replicate rm --all --force region2Replication/bucket-1
mc admin bucket remote rm --arn "$(mc admin bucket remote ls region2Replication/bucket-1 --json | ./jq -r '.RemoteARN')" region2Replication/bucket-1
```

Still unable to remove the region2/bucket-1

**Comment by harshavardhana** [comment]
> mc mb -l region1/bucket-1
> mc mb -l region2/bucket-1

If you create a bucket in a locked state i.e  `mc mb -l` @krisdaniels you are not allowed to delete such buckets. This is as per AWS S3 requirement - you just want to enable versioning for replication alone you should just do `mc version enable`

**Comment by krisdaniels** [comment]
@harshavardhana 
Then why is it possible to delete the first bucket (eventually), while the other one becomes impossible to delete.
Removing the replication first also works. 
However when you do not remove the replication first you end up with a bucket that becomes impossible to delete

**Comment by harshavardhana** [comment]
> @harshavardhana 
> Then why is it possible to delete the first bucket (eventually), while the other one becomes impossible to delete.
> Removing the replication first also works. 
> However when you do not remove the replication first you end up with a bucket that becomes impossible to delete

Will try to see what this is locally..

**Comment by poornas** [comment]
In an active-active replication, when you delete a bucket, all the version deletes in the bucket need to get synced to the other cluster first before bucket deletion can proceed. Since this cannot be guaranteed to be immediate(e.g. other cluster could be temporarily down), the versions are marked deleted and pending purge until the sync is complete to region2/bucket1. If you have a lot of objects in your cluster OR if cluster2 is temporarily down, this will take a while before `mc rb` can succeed.
If you do a `mc stat region1/bucket/objname --versions --debug`, you would see `X-Minio-Replication-Delete-Status` in `PENDING` or `FAILED` status indicating it is waiting to reconcile this delete to the other cluster. As the scanner runs, any such failed or pending reconciliations will be synced, and deleting the version from disk. The bucket deletion will then succeed after all versions are cleared.

**Comment by krisdaniels** [comment]
Hi, waited a few days but the buckets are still not removeable, still same error, any suggestions on how to remove them ?
(other then stopping the cluster and removing the folders from all disks manually)

**Reply by poornas** [reply]
Just do an `mc rm --r --force --versions region1/bucket1` and `mc rm --r --force --versions region2/bucket1` - this should clear the remaining versions. `mc ls --r --versions ` should show no entries after this, allowing you to then delete the bucket

**Reply by krisdaniels** [reply]
it tries to delete some files but ends with (with --debug): 
```
mc: <ERROR> Failed to remove `region1/bucket1` recursively. The specified method is not allowed against this resource.
 (1) rm-main.go:434 cmd.listAndRemove(..) Tags: [region1/bucket1]
 (0) client-s3.go:1199 cmd.(*S3Client).Remove.func1(..)`
```

**Reply by vadmeste** [reply]
This looks like an issue, could you open a new issue with the exact mc command and mc/MinIO version in it ?
